planar/planar.go

```diff
@@ -0,0 +1,86 @@
+package planar
+
+import (
+	"math"
+	"math/big"
+)
+
+const Rad = math.Pi / 180
+
+type PointLineDistanceFunc func(line [2][2]float64, point [2]float64) float64
+
+// PerpendicularDistance  provides the distance between a line and a point in Euclidean space.
+func PerpendicularDistance(line [2][2]float64, point [2]float64) float64 {
+
+	deltaX := line[1][0] - line[0][0]
+	deltaY := line[1][1] - line[0][1]
+	denom := math.Abs((deltaY * point[0]) - (deltaX * point[1]) + (line[1][0] * line[0][1]) - (line[1][1] * line[0][0]))
+	num := math.Sqrt(math.Pow(deltaY, 2) + math.Pow(deltaX, 2))
+	if num == 0 {
+		return 0
+	}
+	return denom / num
+}
+
+type Haversine struct {
+	Semimajor *float64
+}
+
+// PerpendicularDistance returns the distance between a point and a line in meters, using the Harversine distance.
+func (hs Haversine) PerpendicularDistance(line [2][2]float64, point [2]float64) float64 {
+	dist := hs.Distance(line[0], point)
+	angle := Angle3Pts(line[0], line[1], point)
+	return math.Abs(dist * math.Sin(angle))
+}
+
+// Distance returns the Haversine distance between two points in meters
+func (hs Haversine) Distance(pt1, pt2 [2]float64) float64 {
+	rpt1x, rpt2x := pt1[0]*Rad, pt2[0]*Rad
+	distx := rpt1x - rpt2x
+	disty := (pt1[1] * Rad) - (pt2[1] * Rad)
+	dist := 2 * math.Asin(
+		math.Sqrt(
+
+			(math.Pow(math.Sin(distx/2), 2)+math.Cos(rpt1x))*
+				math.Cos(rpt2x)*
+				(math.Pow(math.Sin(disty/2), 2)),
+		),
+	)
```

Comment:
FYI This is [the hoot haversine implementation ](https://github.com/ngageoint/hootenanny/blob/97187cd6f1fe407ff4a23a1b75454ac3a0b0a09a/hoot-core/src/main/cpp/hoot/core/util/GeometryUtils.cpp#L86) I mentioned previously. Looks like I was getting about 9cm precision using float64 with Haversine (likely plenty for your needs).

There are differences in the implementation, but there are a bunch of ways to get to the same answer and I'm sure your tests will reveal it if there are any problems.

Reply:
I don't think I can use hootenanny's implementation as it's GPL'd.